### PR TITLE
Eng 7592  support multiple suggestions

### PR DESCRIPTION
**Pull request by melisa-qordoba**
[ENG-7592]

### Suggestions filtering

This step assumes using suggestions ranking.
Matches in `match_dict` with multiple options can be of 2 types:

I. grammar / syntactic variants ex. people / person, a / an

ex. 
- (A) `Foo a fo fo.` 
- (B) `Foo an fo fo.`

If A is ranked higher, you can hide B (most probably incorrect) by setting `filter_suggestions` to `True`:

```python
lm_path='/path/to/your/kenlm/model.bin'
rmatcher = ReplaceMatcher(nlp, match_dict=match_dict, lm_path=lm_path, filter_suggestions=True)
```

ReplaCy does guessing for you. Expect suggestions filtering if:
-- variants have the same lemma ex. kid / kids, walk / walked
-- variants are DET, ex. a / an
-- variants consist of more than one word, ex. think / think of

II. lexical variants ex. tall / big / huge

ex. 
- (A) `Foo foo big foo.` 
- (B) `Foo foo huge foo.`
- (C) `Foo foo tall foo.`

Set `default_max_count` to any integer to display top n suggestions, ex. `default_max_count=2` would display just A and B.

```python
lm_path='/path/to/your/kenlm/model.bin'
rmatcher = ReplaceMatcher(nlp, match_dict=match_dict, lm_path=lm_path, filter_suggestions=True, default_max_count=2)
```

Additionally, any suggestion item can have `MAX_COUNT` property, which overwrites the above rules ( `filter_suggestions` and `default_max_count` can be omitted ), ex.

```
        "suggestions": [
            [
                {
                    "TEXT": {"IN": ["big", "huge", "tall", "high"]},
                    "MAX_COUNT": 3
                }
            ]
        ],
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -332,6 +332,57 @@ lm_path='/path/to/your/kenlm/model.bin'
...
+
+```python
+lm_path='/path/to/your/kenlm/model.bin'
+rmatcher = ReplaceMatcher(nlp, match_dict=match_dict, lm_path=lm_path, filter_suggestions=True)
+```
+
+ReplaCy does guessing for you. Expect suggestions filtering if:
+-- variants have the same lemma ex. kid / kids, walk / walked
+-- variants are DET, ex. a / an
+-- variants consist of more than one word, ex. think / think of
```
[the remark this answers is not in the document]
Comment (sam-qordoba) [reply README.md:356]: I think I would err on the side of having replaCy assume less, and us having to do more work in the match_dicts to upgrade

File: replacy/suggestion.py [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+        if len(item_options) > 1:
+
+            # E. contains words of the same lemma
+            # ex. [slow, slowly]
+            lemmas = set([])
+            for option in item_options:
+                option_lemmas = set(self.inflector.get_lemmas(option))
+                if len(lemmas & option_lemmas):
+                    return 1
+                lemmas = lemmas.union(option_lemmas)
```
Comment (sam-qordoba) [inline replacy/suggestion.py:134]: not a recommendation, but fun fact: this can be written `lemmas |= option_lemmas`

**Review by sam-qordoba — APPROVED** [review]
Overall: Looks good, let's get it out ASAP, as always.

My biggest hesitation is with Criteria C for automatically assuming a suggestion should be filtered. I feel like "multi-token" is too low-precision. I'd rather have replaCy assume less, and have us do more work (adding MAX_COUNT to more matches) when we update the match_dicts.

I also added some suggestions for using Python built-ins instead of set operations, those aren't must-haves, but I thought we should at least consider them.

I like that this moves some stuff out of `__init__` which was too big.

Finally, in the tests, I think we should have at least one more case, where we explicitly set MAX_COUNT

File: replacy/suggestion.py (hunk 2 above)
Comment (melisa-qordoba) [reply replacy/suggestion.py:134]: :chefkiss:

File: tests/test_inflector.py [hunk 3]
```diff
@@ -31,3 +31,110 @@ def test_inflector(data):
...
+
+"""
+Test lemmatization. 
+Plural and singular forms of nouns should have common (or at least intersecting) sets of lemmas
+Important for max count estimation (see: suggestion.py).
+
+Exceptions to handle separately:
+    {
+        "plural":"person", 
+        "singular": "people"
```
[the remark this answers is not in the document]
Comment (melisa-qordoba) [reply tests/test_inflector.py:44]: Good catch! :+1:

File: tests/test_inflector.py [hunk 4]
```diff
@@ -31,3 +31,110 @@ def test_inflector(data):
...
+        "singular": "quails"
+    }
+]
+
+@pytest.mark.parametrize("pair", irregular_nouns)
+def test_lemmatization(pair):
+    singular_lemmas = set(inflector.get_lemmas(pair["singular"]))
+    plural_lemmas = set(inflector.get_lemmas(pair["plural"]))
+
+    assert len(singular_lemmas & plural_lemmas) > 0, "lemmas are different!"
```
[the remark this answers is not in the document]
Comment (melisa-qordoba) [reply tests/test_inflector.py:140]: This is different for people / person and on / oxen.
But you are right - added xfail cases to the test.
Now, whenever we upgrade lemminflect, we can check whether lemmatization of irregular words is the same.

File: replacy/suggestion.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+
+            # if non empty (can be when matching with OP)
+            if len(cased_options):
+                suggestion_variant = SuggestionVariants(cased_options, max_count)
+                suggestions.append(suggestion_variant)
+
+        return suggestions
+
+
+class SuggestionVariants:
```
[the remark this answers is not in the document]
Comment (melisa-qordoba) [reply replacy/suggestion.py:282]: :smiley:

File: tests/test_max_count.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
+import pytest
+import spacy
+
+from replacy import ReplaceMatcher
+
+nlp = spacy.load("en_core_web_sm")
+
+# They read us the stories they themselves had written.
+
+match_dict = {
```
[the remark this answers is not in the document]
Comment (melisa-qordoba) [reply tests/test_max_count.py:10]: done :+1:

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (melisa-qordoba) [reply README.md:356]: Hmm I would rewrite that match to:
```
{"TEXT": "who is"},
{"TEXT": ["single", "unmarried"]}
```
Then those rules would be ok :smile: 
But yeah, I agree suppression might be confusing, and some matches will need rewrites.
I promise to:
- add suppression warnings
- check `match_dicts` in dependent services
- rethink how to make those criteria better.

In other words... let's :shipit: , then fix the reality :rofl:

File: replacy/suggestion.py [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+        # ex. []
+        if not len(item_options):
+            return 1
+
+        # B. contains non letters
+        # ex. ["", ","]
+        if not all([o.isalpha() for o in item_options]):
+            return 1
+
+        # C. is multi token
```
[the remark this answers is not in the document]
Comment (melisa-qordoba) [reply replacy/suggestion.py:110]: Same applies here. Done for fast :shipit: Let's make it better ... later :sweat_smile:

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (melisa-qordoba) [reply README.md:356]: added simple logging :sweat_smile:

**Pull request merged**